components/authz-service/storage/postgres/migration/sql/61_use_role_db_id_in_iam_statements.up.sql

```diff
@@ -0,0 +1,233 @@
+ALTER TABLE iam_statements ALTER CONSTRAINT iam_statements_role_id_fkey DEFERRABLE INITIALLY DEFERRED;
+
+-- i wanted use the above FKEY constraint with ON DELETE to clean up any statement where the
+-- role getting removed would result in a NULL role_id in iam_statements. that is doable, but
+-- what becomes tricky is adding the additional logic to completely remove the statement if the affected
+-- policy would also have no actions. so instead of splitting that logic up, let's handle it all in a trigger.
+
+CREATE OR REPLACE FUNCTION purge_statements_with_no_actions_or_role() RETURNS TRIGGER AS $$
+  BEGIN
+    -- for statements that will still have actions, simply remove the role since those
+    -- statements are still valid
+    UPDATE iam_statements SET role_id=NULL WHERE role_id=OLD.db_id AND actions != '{}';
+
+    -- for statements that become invalid (no role or actions), delete them
+    DELETE FROM iam_statements WHERE role_id=OLD.db_id AND actions = '{}';
+
+    -- if as a result, a policy now has no statements, remove the policy unless it's
+    -- chef-managed (chef-managed case should never happen since chef-managed roles can't be
+    -- deleted, but just to be safe adding it in)
+    DELETE FROM iam_policies USING iam_policies AS p
+      LEFT OUTER JOIN iam_statements AS s ON p.db_id=s.policy_id
+      WHERE s.policy_id IS NULL AND iam_policies.id=p.id AND p.type != 'chef-managed';
+
+    RETURN NULL;
+  END
+$$
+LANGUAGE plpgsql;
+
+CREATE TRIGGER on_role_deletion AFTER DELETE ON iam_roles
+FOR EACH ROW
+EXECUTE PROCEDURE purge_statements_with_no_actions_or_role();
+
+CREATE OR REPLACE FUNCTION role_db_id (_id TEXT)
+    RETURNS INTEGER
+    AS $$
+    SELECT
+        db_id
+    FROM
+        iam_roles
+    WHERE
+        id = _id;
+$$
+LANGUAGE SQL;
+
+CREATE OR REPLACE FUNCTION query_policy (_policy_id TEXT, _projects_filter TEXT[])
+    RETURNS json
+    AS $$
+    WITH temp AS (
+        SELECT
+            pol.db_id,
+            pol.id,
+            pol.name,
+            pol.type,
+            -- get policy's statements using temporary table
+            ( WITH statement_rows AS (
+                    SELECT
+                        stmt.db_id,
+                        stmt.id,
+                        stmt.effect,
+                        stmt.actions,
+                        stmt.resources,
+                        (
+                          SELECT COALESCE((SELECT id FROM iam_roles WHERE db_id=stmt.role_id), '') AS role
+                        ),
+                        (
+                            SELECT
+                                COALESCE(json_agg(proj.id) FILTER (WHERE proj.id IS NOT NULL),
+                                    '[]')
+                                FROM iam_statement_projects AS stmt_projs
+                            LEFT OUTER JOIN iam_projects AS proj ON stmt_projs.statement_id = stmt.db_id WHERE stmt_projs.project_id = proj.db_id) AS projects FROM iam_statements stmt
```

Comment:
Indents in this function make for very difficult reading (inherited, I realize). But at least these couple changes will help a bit.

```suggestion
                            LEFT OUTER JOIN iam_projects AS proj
                            ON stmt_projs.statement_id = stmt.db_id
                            WHERE stmt_projs.project_id = proj.db_id
                        ) AS projects 
                    FROM iam_statements stmt
```